EIPS/eip-2579.md

```diff
@@ -0,0 +1,128 @@
+---
+eip: 2579
+title: Merkle Tree Generalized Index
+author: naxe (ngmachado @LogicB0x)
+discussions-to: https://ethereum-magicians.org/t/merkle-tree-generalized-index/4162/
+status: Draft
+type: Standards Track
+category: ERC
+created: 2020-03-31
+---
+
+## Simple Summary
+
+Using a Merkle Tree with a Generalized Index coordination system as used in other systems with the necessary adaptation to a smart contract environment, reducing the computation requirement and storage space.
+
+## Motivation
+
+Bring the capacity to the developer community to implement a selective proof or multi proof system “as a service” in the form of a library, helping the intercommunication between different ecosystems in Ethereum space by using the same library to process profs.
+
+## Specification
+
+### Definitions:
+
+#### The tree must have the following properties:
+
+The tree is defined as a Full Complete Binary Tree.
+
+There is a one-way function that is easy to compute, but difficult to invert, ```hash(x)```.
+
+For each node and leaves, there is a unique positive integer called generalized index, ```gi```.
+
+The index is attributed for each node from top to bottom, left to right so that parent generalized index, ```parent=⌊(gi-1) / 2⌋ ```, left child ```left = 2gi + 1```,  right child ```right = 2gi + 2``` is computable to all cases.
+
+Each node is composed of the hash of the two immediate children by, ```node = hash (hash( left child ) || hash( right child))```
+
+Each leaf is composed of the hash of some proof such that  ```leaf = hash(proof)```.
+
+Each leaf there is a sibling leaf, sibling = hash(proof), such that parent node can be computed as ```parent = hash(leaf || sibling)```
+
+#### Attribution of index:
+
+Starting at 0 from the root node, increment by 1 for each subnode found, following the direction top to bottom and left to right.
+```
+     1   
+  2     3
+4  5   6  7
+     ...
+```
+
+The set of indexes are grouped in opposition of attribution, **bottom up, left to right**.
+
+We called this set the Generalized Index Operation Order ```giop```.
+
+#### Construction of proof:
+
+Each node or leaf are given in the form of ```hash(proof)```.
+
+For each verifier path is given the leaf and also the sibling leaf.
+
+For each pair of leaf and sibling is given the generalized index that must be of the leaf and not the sibling.
+
+For each tree path not in the construction proof set, is given the additional authentication information that minimizes the computation.
+
+#### Reserved Word
+
+All words are accepted with the exception of ```hash(0)```, called **BLANK POINT**.
```

Comment:
I'm not sure I follow what you mean here by "reserved word"?

Reply:
Sorry if is not clear.

The reserved word is just to exclude **sha3(0)** from the acceptable input set. This is used to signal in the same array the diferences between the proofs (leaves) and nodes (intermediary nodes of the tree)